agent/taskresource/cgroup/cgroup.go

```diff
@@ -44,143 +44,230 @@ var (
 		control:         control,
 		cgroupRoot:      cgroupRoot,
 		cgroupMountPath: cgroupMountPath,
 		resourceSpec:    resourceSpec,
 		ioutil:          ioutilwrapper.NewIOUtil(),
 	}
+	c.initializeResourceStatusToTransitionFunction()
+	return c
+}
+
+func (cgroup *CgroupResource) initializeResourceStatusToTransitionFunction() {
+	resourceStatusToTransitionFunction := map[taskresource.ResourceStatus]func() error{
+		taskresource.ResourceStatus(CgroupCreated): cgroup.Create,
+	}
+	cgroup.resourceStatusToTransitionFunction = resourceStatusToTransitionFunction
 }
 
 // SetDesiredStatus safely sets the desired status of the resource
-func (c *CgroupResource) SetDesiredStatus(status taskresource.ResourceStatus) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
+func (cgroup *CgroupResource) SetDesiredStatus(status taskresource.ResourceStatus) {
+	cgroup.lock.Lock()
+	defer cgroup.lock.Unlock()
 
-	c.desiredStatusUnsafe = status
+	cgroup.desiredStatusUnsafe = status
 }
 
 // GetDesiredStatus safely returns the desired status of the task
-func (c *CgroupResource) GetDesiredStatus() taskresource.ResourceStatus {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
+func (cgroup *CgroupResource) GetDesiredStatus() taskresource.ResourceStatus {
+	cgroup.lock.RLock()
+	defer cgroup.lock.RUnlock()
 
-	return c.desiredStatusUnsafe
+	return cgroup.desiredStatusUnsafe
 }
 
 // GetName safely returns the name of the resource
-func (c *CgroupResource) GetName() string {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
+func (cgroup *CgroupResource) GetName() string {
+	cgroup.lock.RLock()
+	defer cgroup.lock.RUnlock()
 
 	return resourceName
 }
 
+// DesiredTerminal returns true if the cgroup's desired status is REMOVED
+func (cgroup *CgroupResource) DesiredTerminal() bool {
+	cgroup.lock.RLock()
+	defer cgroup.lock.RUnlock()
+
+	return cgroup.desiredStatusUnsafe == taskresource.ResourceStatus(CgroupRemoved)
+}
+
+// KnownCreated returns true if the cgroup's known status is CREATED
+func (cgroup *CgroupResource) KnownCreated() bool {
+	cgroup.lock.RLock()
+	defer cgroup.lock.RUnlock()
+
+	return cgroup.knownStatusUnsafe == taskresource.ResourceStatus(CgroupCreated)
+}
+
+// TerminalStatus returns the last transition state of cgroup
+func (cgroup *CgroupResource) TerminalStatus() taskresource.ResourceStatus {
+	return taskresource.ResourceStatus(CgroupRemoved)
+}
+
+// GetNextKnownStateProgression returns the state that the resource should
+// progress to based on its `KnownState`.
+func (cgroup *CgroupResource) GetNextKnownStateProgression() taskresource.ResourceStatus {
+	return cgroup.GetKnownStatus() + 1
+}
+
+// ApplyTransition calls the function required to move to the specified status
+func (cgroup *CgroupResource) ApplyTransition(nextState taskresource.ResourceStatus) (bool, error) {
+	transitionFunc, ok := cgroup.resourceStatusToTransitionFunction[nextState]
+	if !ok {
+		seelog.Criticalf("Cgroup Resource [%s]: unsupported desired state transition [%s]: %s",
```

Comment:
This shouldn't be a `Critical` log line either. `Error` seems like a more appropriate level